File: VPNKit/VPNManager.m
//
//  VPNManager.m
//  VPN
//
//  Created by ZhaoWei on 15/6/9.
//  Copyright (c) 2015年 csdept. All rights reserved.
//

#import "VPNManager.h"
#import "VPN.h"
#import "VPNKeyChainWrapper.h"
#import "VPN+OnDemandRules.h"
#import "UIDevice+Type.h"

static NSString * const kAppGroupIdentifier = @"group.com.zte.VPN";

@interface VPNManager ()

@property (nonatomic, strong) NEVPNManager   *manager;
@property (nonatomic, strong) NSUserDefaults *userDefault;

@end

@implementation VPNManager

+ (VPNManager *)sharedInstance
{
    static VPNManager *sharedInstance;
    
    static dispatch_once_t onceToken;
    dispatch_once(&onceToken, ^{
        sharedInstance = [[VPNManager alloc] init];
    });
    
    return sharedInstance;
}

- (instancetype)init
{
    self = [super init];
    if (self)
    {
        _manager = [NEVPNManager sharedManager];
        
        _userDefault = [[NSUserDefaults alloc] initWithSuiteName:kAppGroupIdentifier];
    }
    return self;
}

- (NEVPNStatus)status
{
    return self.manager.connection.status;
}

- (void)loadFromPreferencesWithCompletionHandler:(void(^)(NSError *error))completionHandler
{
    [self.manager loadFromPreferencesWithCompletionHandler:^(NSError *error) {
        if (error)
        {
            NSLog(@"LoadFromPreferences error: %@", error);
        }
        
        if (completionHandler)
        {
            completionHandler(error);
        }
    }];
}

- (void)removeFromPreferencesWithCompletionHandler:(void(^)(NSError *error))completionHandler
{
    [self.manager removeFromPreferencesWithCompletionHandler:^(NSError *error) {
        if (error)
        {
            NSLog(@"RemoveFromPreferences error: %@", error);
        }
        
        if (completionHandler)
        {
            completionHandler(error);
        }
    }];
}

- (void)connectVPN:(VPN *)vpn titlePrefix:(NSString *)prefix completionHandler:(void(^)(NSError *error))completionHandler
{
    [self.manager loadFromPreferencesWithCompletionHandler:^(NSError *error) {
        if (error)
        {
            NSLog(@"LoadFromPreferences error: %@", error);

            if (completionHandler)
            {
                completionHandler(error);
            }
        }
        else
        {
            [self saveAndStartVPNTunnel:vpn titlePrefix:prefix completionHandler:completionHandler];
        }
    }];
}

- (void)saveAndStartVPNTunnel:(VPN *)vpn titlePrefix:(NSString *)prefix completionHandler:(void(^)(NSError *error))completionHandler
{
    NEVPNProtocolIPSec *p = [[NEVPNProtocolIPSec alloc] init];
    p.authenticationMethod = NEVPNIKEAuthenticationMethodSharedSecret;
    p.useExtendedAuthentication = YES;
    p.serverAddress = vpn.server;
    p.username = vpn.account;
    p.localIdentifier = vpn.groupName;
    p.disconnectOnSleep = [vpn.disconnectOnSleep boolValue];
    p.passwordReference = [[VPNKeyChainWrapper sharedInstance] passwordForVPNID:vpn.VPNID];
    p.sharedSecretReference = [[VPNKeyChainWrapper sharedInstance] secretForVPNID:vpn.VPNID];
    
    //iOS 9使用protocolConfiguration，代替protocol
    if ([[UIDevice currentDevice] isGEVersion:@"9"])
    {
        self.manager.protocolConfiguration = p;
    }
    else
    {
        self.manager.protocol = p;
    }
    
    self.manager.enabled = YES;
    
    if (![vpn.isOnDemand boolValue])
    {
        self.manager.onDemandEnabled = NO;
    }
    else
    {
        self.manager.onDemandEnabled = YES;
        NEEvaluateConnectionRule *connectRule = [[NEEvaluateConnectionRule alloc] initWithMatchDomains:[vpn onDemandRules] andAction:NEEvaluateConnectionRuleActionConnectIfNeeded];
        NEOnDemandRuleEvaluateConnection *rule = [NEOnDemandRuleEvaluateConnection new];
        rule.connectionRules = @[connectRule];
        self.manager.onDemandRules = @[rule];
    }
    
    NSString *displayName = [[NSBundle mainBundle] objectForInfoDictionaryKey:@"CFBundleDisplayName"];
    NSString *description;
    if (prefix)
    {
        description = [NSString stringWithFormat:@"%@-%@-%@", displayName, prefix, vpn.title];
    }
    else
    {
        description = [NSString stringWithFormat:@"%@-%@", displayName, vpn.title];
    }
    self.manager.localizedDescription = description;
    
    [self.manager saveToPreferencesWithCompletionHandler:^(NSError *error) {
        if (error)
        {
            NSLog(@"SaveToPreferences error: %@", error);
        
            if (completionHandler)
            {
                completionHandler(error);
            }
        }
        else
        {
            NSError *startError;
            [self.manager.connection startVPNTunnelAndReturnError:&startError];
            
            if (startError)
            {
                NSLog(@"StartVPNTunnel error: %@", startError);
                
                //iOS 9第一次saveToPreferences安装VPN到设备之后回到App，需要再调用loadFromPreferences，加载VPN设置
                if (startError.domain == NEVPNErrorDomain && startError.code == NEVPNErrorConfigurationInvalid)
                {
                    dispatch_async(dispatch_get_main_queue(), ^{
                        [[VPNManager sharedInstance] loadFromPreferencesWithCompletionHandler:nil];
                    });
                }
                
                if (completionHandler)
                {
                    completionHandler(startError);
                }
            }
            else
            {
                if (completionHandler)
                {
                    completionHandler(nil);
                }
            }
        }
    }];
}

- (void)disConnect
{
    [self.manager.connection stopVPNTunnel];
}

#pragma mark - NSUserDefault
- (void)setObject:(id)object forKey:(id)key
{
    [self.userDefault setObject:object forKey:key];
    [self.userDefault synchronize];
}

- (NSString *)stringForKey:(NSString *)key
{
    return [self.userDefault stringForKey:key];
}

- (void)removeObjectForKey:(NSString *)key
{
    [self.userDefault removeObjectForKey:key];
}

@end

File: VPNManager/VPNManager.swift
//
//  VPNManager.swift
//  VPN
//
//  Created by CP3 on 15/10/31.
//  Copyright © 2015年 CP3. All rights reserved.
//

import UIKit
import NetworkExtension
import VPNKit

typealias CompletionHandler = ((NSError?) -> Void)

class VPNManager: NSObject {
    
    static let sharedInstance = VPNManager()
    
    var manager = NEVPNManager.sharedManager()
    
    /// VPN connection status.
    var status: NEVPNStatus {
        return manager.connection.status
    }
    
    /**
     This function loads the current VPN configuration from the caller's VPN preferences.
     - parameter completionHandler: A block that will be called on the main thread when the load operation is completed
     */
    func loadVPNPreferences(completionHandler: CompletionHandler? = nil) {
        
        manager.loadFromPreferencesWithCompletionHandler {
            (error: NSError?) -> Void in
            
            if let loadError = error {
                print("Load VPN preferences error = \(loadError)")
            }
            
            if let completion = completionHandler {
                completion(error)
            }
        }
    }
    
    /**
     This function removes the VPN configuration from the caller's VPN preferences.
     - parameter completionHandler: A block that will be called on the main thread when the remove operation is completed
     */
    func removeVPNPreferences(completionHandler: CompletionHandler? = nil) {
        
        manager.removeFromPreferencesWithCompletionHandler { (error: NSError?) -> Void in
            
            if let removeError = error {
                print("Remove VPN preferences error = \(removeError)")
            }
            
            if let completion = completionHandler {
                completion(error)
            }
        }
    }
    
    /**
     Connecting VPN.
     
     - parameter vpn:               The connecting VPN.
     - parameter prefix:            Distinguish App from App Extension
     - parameter completionHandler: A block that will be called on the main thread when the connect operation is completed
     */
    func connectVPN(vpn: VPN, titlePrefix prefix: String? = nil, completionHandler: CompletionHandler? = nil) {
        
        self.saveVPN(vpn, titlePrefix: prefix) { [unowned self] (error: NSError?) -> Void in
            
            //Save preferences error
            if let saveError = error {
                if let completion = completionHandler {
                    completion(saveError)
                }
            } else {
                //startVPNTunnel
                self.startConnect(completionHandler)
            }
        }
    }
    
    /**
     Save VPN preferences.
     
     - parameter vpn:               The connecting VPN.
     - parameter prefix:            Distinguish App from App Extension
     - parameter completionHandler: A block that will be called on the main thread when the save operation is completed
     */
    func saveVPN(vpn: VPN, titlePrefix prefix: String?, completionHandler: CompletionHandler? = nil) {
        
        self.loadVPNPreferences { [unowned self] (error: NSError?) -> Void in
            
            //Load preferences error
            if let loadError = error {
                if let completion = completionHandler {
                    completion(loadError)
                }
            } else {
                
                //Save preferences
                let ipsec = NEVPNProtocolIPSec()
                ipsec.authenticationMethod = .SharedSecret
                ipsec.useExtendedAuthentication = true
                ipsec.serverAddress = vpn.server
                ipsec.username = vpn.account
                ipsec.localIdentifier = vpn.groupName
                ipsec.disconnectOnSleep = vpn.disconnectOnSleep
                ipsec.passwordReference = VPNKeyChainManager.sharedInstance.passwordForVPNID(vpn.VPNID)
                ipsec.sharedSecretReference = VPNKeyChainManager.sharedInstance.secretForVPNID(vpn.VPNID)
                
                if #available(iOS 9.0, *) {
                    self.manager.protocolConfiguration = ipsec
                } else {
                    self.manager.`protocol` = ipsec
                }
                
                self.manager.enabled = true
                
                if vpn.isOnDemand {
                    self.manager.onDemandEnabled = true
                    
                    let connectRule = NEEvaluateConnectionRule(matchDomains: vpn.onDemandRules(), andAction: .ConnectIfNeeded)
                    let rule = NEOnDemandRuleEvaluateConnection()
                    rule.connectionRules = [connectRule]
                    self.manager.onDemandRules = [rule]
                } else {
                    self.manager.onDemandEnabled = false
                }
                
                let displayName = NSBundle.mainBundle().objectForInfoDictionaryKey("CFBundleDisplayName") as! String
                let title = vpn.title ?? ""
                var description: String
                if let descPrefix = prefix {
                    description = displayName + "-" + descPrefix + "-" + title
                } else {
                    description = displayName + "-" + title
                }
                self.manager.localizedDescription = description
                
                self.manager.saveToPreferencesWithCompletionHandler { (error: NSError?) -> Void in
                    
                    if let saveError = error {
                        print("Save VPN preferences error = \(saveError)")
                    }
                    
                    if let completion = completionHandler {
                        completion(error)
                    }
                }
            }
        }
    }
    
    /**
     Connect VPN according to the current VPN preferences. First call saveVPN method before it.
     
     - parameter completionHandler: A block that will be called on the main thread when the start operation is completed
     */
    func startConnect(completionHandler: CompletionHandler? = nil) {
        do {
            try self.manager.connection.startVPNTunnel()
            if let completion = completionHandler {
                completion(nil)
            }
        } catch let catchError as NSError {
            print("VPN start VPN tunnel error = \(catchError)")
            
            //iOS 9第一次saveToPreferences安装VPN到设备，需要再调用loadFromPreferences，加载VPN设置
            if (catchError.domain == NEVPNErrorDomain && catchError.code == NEVPNError.ConfigurationInvalid.rawValue)
            {
                dispatch_async(dispatch_get_main_queue(), {
                    self.loadVPNPreferences()
                })
            }
            
            if let completion = completionHandler {
                completion(catchError)
            }
        }
    }
    
    /**
     This function is used to stop the VPN tunnel.
     */
    func disConnect() {
        
        manager.connection.stopVPNTunnel()
    }
}

